docs/cookbook.md

```diff
@@ -0,0 +1,87 @@
+# Odoyle's Rules Cookbook
+
+## Initializing values
+
+Suppose we have an aggregate:
+
+```clojure
+(o/ruleset
+  {#_...
+   ::agregate-things
+   [:what 
+    [thing-id :thing/name  thing-name]
+    [thing-id :thing/owner person-id]
+    :then-finally
+    (->> (o/rules o/*session* ::aggregate-things)
+         (group-by :thing-id)
+         (o/insert o/*session* :global :domain/things)
+         (o/reset!))]})
+```
+
+The trouble is that `:thing/owner` may be an attribute of some things, but not all of them. We want to aggregate all the things.
+
+The solution is to use a rule that initializes a `nil` value for `:thing/owner`:
+
+```clojure
+(o/ruleset
+  {#_...
+   ::initialize-thing-owner
+   [:what [_ :thing/id id {:then not=}]
+    :then (o/insert! id :thing/owner nil)]})
```

Comment:
Instead of using `{:then not=}` here you could use `o/contains?` to check if the fact already exists:

```clj
[:what
 [_ :thing/id id]
 :then
 (when-not (o/contains? o/*session* id :thing/owner)
   (o/insert! id :thing/owner nil))]
```

The problem with using `{:then not=}` here is that you may actually need to update `:thing/id` for some reason, and if you do, it'll set the `:thing/owner` to nil even if it already had a value.

Reply:
Interesting! I've run into that limitation with `{:then not=}` and worked around it sometimes when I've needed to. Using `o/contains?` is perfect for this. I'll update the PR.

Reply:
@oakes - one mistake I started making here pretty quickly was to do this:

```clojure
[:what
 [_ :thing/id id]
 :when (not (o/contains? o/*session* id :thing/owner))
 :then
 (o/insert! id :thing/owner nil)]
```

This case can happen, which is strange:

```clojure
[:what
 [id :thing/id id]
 [id :thing/owner owner]
 :when (not (o/contains? o/*session* id :thing/owner))
 :then
 (o/insert! id :thing/owner nil)]
```

In this case, there is a match, which is counter-intuitive to me.

Reply:
@oakes - one mistake I started making here pretty quickly was to do this:

```clojure
[:what
 [_ :thing/id id]
 :when (not (o/contains? o/*session* id :thing/owner))
 :then
 (o/insert! id :thing/owner nil)]
```

This case can happen, which is strange:

```clojure
[:what
 [id :thing/id id]
 [id :thing/owner owner]
 :when (not (o/contains? o/*session* id :thing/owner))
 :then
 (o/insert! id :thing/owner nil)]
```

In this case, there is a match, which is counter-intuitive to me.

Reply:
I believe *session* isn't bound to anything in :when blocks. I need to fix that. I'm about to board a plane but I'll look into that soon.

Reply:
If you can, try the latest commit. It should fix this problem.

Reply:
Did you get a chance to try the latest commit and see if it fixes this?